src/System.Net.Sockets/src/System/Net/Sockets/OverlappedAsyncResult.Windows.cs

```diff
@@ -42,11 +41,11 @@ internal unsafe int GetSocketAddressSize()
         // These calls are outside the runtime and are unmanaged code, so we need
         // to prepare specific structures and ints that lie in unmanaged memory
         // since the overlapped calls may complete asynchronously.
-        internal void SetUnmanagedStructures(byte[] buffer, int offset, int size, Internals.SocketAddress socketAddress, bool pinSocketAddress)
+        internal void SetUnmanagedStructures(byte[] buffer, Internals.SocketAddress socketAddress)
         {
             // Fill in Buffer Array structure that will be used for our send/recv Buffer
             _socketAddress = socketAddress;
-            if (pinSocketAddress && _socketAddress != null)
+            if (_socketAddress != null)
             {
                 object[] objectsToPin = null;
                 objectsToPin = new object[2];
```

Comment:
Which array -- objectsToPin?

SetUnmanagedStructures is only called once per IAsyncResult, so there's not an easy way to reuse stuff like this.

A better answer is to make IAsyncResult more like the approach I've taken in this code with SAEA: don't use Overlapped's pinning stuff (which forces the objectsToPin allocation), instead use explicit GCHandles.  That way we can have exactly the # of GCHandles we need for a particular operation.  (Well not exactly, since the # of buffers can vary, but we can do a lot better than we are today.)

This is a fairly involved change, so I didn't tackle it for now.